### PR TITLE
Added named tuple sample shapes to all formats

### PR DESCRIPTION
**Pull request by cczhu**
Added named tuple sample shapes to all format payloads and stream readers.  Created a `sample_shape` property in `vlbi_base` that's inherited by all formats to generate a squeezed sample shape as needed.  `squeeze` is now a keyword for stream reader `__init__`, which adds it as an instance property so that both `sample_shape` and `read` have access to it.  Left stream writers alone so that the user can write squeezed and unsqueezed data consecutively without having to change instance attributes.  All references to `nthread` or `nchan` now redirect to sample shape fields, allowing us to be self-consistent with the definition of `nthread`.  For formats where `nthread` used to be `len(thread_ids)` when `thread_ids` is user-defined and less than the total number of threads, now `sample_shape` takes this into account directly in `nchan` (leading to a mismatch between that value and `header.nchan`, but I think it's an acceptable loss).

User-selected threads/channels are still always called `thread_ids`, which feels wrong for Mark 4, DADA, etc.

Annoyingly DADA has a `sample_shape` header derived property.  I feel like we should either remove it, or include one for all other headers and have it passed to their payloads.  I think the former makes much more sense.

### CODE REVIEW

File: baseband/vlbi_base/base.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import io
 import warnings
 import numpy as np
+from collections import namedtuple
```
Comment (mhvk) [inline baseband/vlbi_base/base.py]: remove

File: baseband/vdif/frame.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
                         unicode_literals)
 
 import numpy as np
+from collections import namedtuple
```
Comment (mhvk) [inline baseband/vdif/frame.py]: remove

File: baseband/dada/base.py [hunk 3]
```diff
@@ -210,9 +212,9 @@ def _frame_info(self):
         return divmod(self.offset, self.samples_per_frame)
 
     def _unsqueeze(self, data):
```
Comment (mhvk) [inline baseband/dada/base.py]: Useful for all?

File: baseband/dada/payload.py [hunk 4]
```diff
@@ -93,3 +97,20 @@ def fromfile(cls, fh, header=None, memmap=False, payloadsize=None,
...
+        ----------
+        data : `numpy.ndarray`
+            Data to be encoded. The last dimension is taken as the number of
+            channels.
+        bps : int
+            Number of bits per sample to use (for complex data, for real and
+            imaginary part separately; default: 2).
+        """
+        payload = super(DADAPayload, cls).fromdata(data, bps=bps)
+        payload.sample_shape = cls._sample_shape_cls(*data.shape[1:])
```
Comment (mhvk) [inline baseband/dada/payload.py]: Move to `vlbi_base` and define `_sample_shape_cls = tuple` in there so that it works even if people do not override `_sample_shape_cls`

File: baseband/gsb/base.py [hunk 5]
```diff
@@ -124,8 +125,17 @@ def __init__(self, fh_ts, fh_raw, header0, thread_ids=None,
                                  (1 if rawdump else len(fh_raw[0])) //
                                  (nchan * (2 if complex_data else 1)))
 
+        # Temporary warning that specific thread reading isn't supported.
+        if not rawdump and len(thread_ids) != len(fh_raw):
+            warnings.warn("Baseband.gsb currently does not support reading"
```
Comment (mhvk) [inline baseband/gsb/base.py]: Make this an exception - if it doesn't work, it doesn't work...

File: baseband/gsb/base.py [hunk 6]
```diff
@@ -124,8 +125,17 @@ def __init__(self, fh_ts, fh_raw, header0, thread_ids=None,
...
+                          " specific threads.  All {0} threads will be "
+                          "read.".format(len(fh_raw)))
+        # sample_shape uses len(fh_raw) instead of len(thread_ids)!
+        sample_shape = GSBPayload._sample_shape_cls_1thread(nchan) if rawdump \
```
Comment (mhvk) [inline baseband/gsb/base.py]: nitpick: write out if statement. Or at least use paretheses rather than `\`

File: baseband/mark4/payload.py [hunk 7]
```diff
@@ -210,6 +211,8 @@ class Mark4Payload(VLBIPayloadBase):
                  (8, 2, 2): decode_8chan_2bit_fanout2,
                  (8, 2, 4): decode_8chan_2bit_fanout4}
 
+    _sample_shape_cls = namedtuple('sample_shape', 'nchan')
```
Comment (mhvk) [inline baseband/mark4/payload.py]: Should we do `namedtuple('SampleShape', 'nchan')`

File: baseband/mark4/tests/test_mark4.py [hunk 8]
```diff
@@ -502,8 +495,44 @@ def test_filestreamer(self, tmpdir):
                 conv_bytes = s.read()
                 assert conv_bytes == orig_bytes
 
-    def test_corrupt_stream(self):
-        with mark4.open(SAMPLE_FILE, 'rb') as fh, io.BytesIO() as s:
+        # Test that squeeze attribute works on read and sample shape, but is
+        # overridden when reading in-place.
+        with mark4.open(SAMPLE_FILE, 'rs', ntrack=64, decade=2010) as fh:
+            assert tuple(fh.sample_shape) == (8,)
```
Comment (mhvk) [inline baseband/mark4/tests/test_mark4.py]: no need to cast. Also good to check
```
assert fh.sample_shape.nchan == 8
```

File: baseband/mark5b/base.py [hunk 9]
```diff
@@ -158,29 +159,34 @@ class Mark5BStreamReader(VLBIStreamReaderBase, Mark5BFileReader):
...
     frames_per_second : int, optional
         Needed to calculate timestamps. If not given, will be inferred from
         ``sample_rate``, or by scanning the file.
     sample_rate : `~astropy.units.Quantity`, optional
         Rate at which each thread is sampled (bandwidth * 2; frequency units).
+    squeeze : bool
+        If `True` (default), remove channel and thread dimensions if unity.
     """
 
     _frame_class = Mark5BFrame
```
Comment (mhvk) [inline baseband/mark5b/base.py]: Maybe add `_payload_class` or use `self._frame_class._payload_class._sample_shape_cls(...)`

Is `_frame_class` even used? If so, shouldn't it be in the binary reader?

Comment (mhvk) [reply baseband/mark5b/base.py]: See separate issue #87

File: baseband/mark5b/base.py [hunk 10]
```diff
@@ -201,11 +207,9 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
             file is read.
         fill_value : float or complex
             Value to use for invalid or missing data.
-        squeeze : bool
-            If `True` (default), remove channel and thread dimensions if unity.
         out : `None` or array
             Array to store the data in. If given, count will be inferred,
-            and squeeze is set to `False`.
+            and squeeze behavior will be overridden to False.
```
Comment (mhvk) [inline baseband/mark5b/base.py]: Keep back ticks.

File: baseband/mark5b/tests/test_mark5b.py [hunk 11]
```diff
@@ -400,10 +396,47 @@ def test_filestreamer(self, tmpdir):
...
+            fh.seek(0)
+            out = np.zeros((12, 8))
+            fh.read(out=out)
+            assert fh.tell() == 12
+            assert np.all(out.squeeze() == record[:12])
+
+        with mark5b.open(SAMPLE_FILE, 'rs', nchan=8, bps=2,
+                         sample_rate=32*u.MHz, ref_mjd=57000,
+                         squeeze=False) as fh:
+            assert tuple(fh.sample_shape) == (8,)
```
Comment (mhvk) [inline baseband/mark5b/tests/test_mark5b.py]: No conversion necessary

Comment (mhvk) [reply baseband/mark5b/tests/test_mark5b.py]: make sure the payload one is done somewhere as well.

File: baseband/vlbi_base/base.py [hunk 12]
```diff
@@ -115,7 +117,8 @@ def _frame_info(self):
 
     def __repr__(self):
         return ("<{s.__class__.__name__} name={s.name} offset={s.offset}\n"
-                "    samples_per_frame={s.samples_per_frame}, nchan={s.nchan},"
+                "    samples_per_frame={s.samples_per_frame},"
+                " nchan={s._sample_shape.nchan},"
```
[the remark this answers is not in the document]
Comment (cczhu) [reply baseband/vlbi_base/base.py]: Done; reorganized the print statement a bit to fit within 79 characters.

File: baseband/vdif/base.py [hunk 13]
```diff
@@ -263,8 +268,9 @@ def _get_time(self, header):
 
     def __repr__(self):
         return ("<{s.__class__.__name__} name={s.name} offset={s.offset}\n"
-                "    nthread={s.nthread}, "
-                "samples_per_frame={s.samples_per_frame}, nchan={s.nchan},\n"
+                "    nthread={s._sample_shape.nthread}, "
```
[the remark this answers is not in the document]
Comment (cczhu) [reply baseband/vdif/base.py]: Still necessary because we want to print out the EDV and station.  I also added a `__repr__` method to `GSBStreamBase` (currently it prints the names of all raw data files in an ugly way...).

File: baseband/vdif/base.py [hunk 14]
```diff
@@ -308,6 +317,7 @@ def __init__(self, fh_raw, thread_ids=None, frames_per_second=None,
         self._framesetsize = fh_raw.tell()
         super(VDIFStreamReader, self).__init__(fh_raw, header, thread_ids,
                                                frames_per_second, sample_rate)
```
[the remark this answers is not in the document]
Comment (cczhu) [reply baseband/vdif/base.py]: I've made some further changes to this - see summary comments below.

File: baseband/vdif/base.py [hunk 15]
```diff
@@ -362,8 +370,10 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
...
                 count = self.size - self.offset
 
-            out = np.empty((self.nthread, count, self.nchan),
+            out = np.empty((self._sample_shape.nthread, count,
+                            self._sample_shape.nchan),
                            dtype=self._frameset.dtype).transpose(1, 0, 2)
+            squeeze = self.squeeze
         else:
             count = out.shape[0]
             squeeze = False
```
[the remark this answers is not in the document]
Comment (cczhu) [reply baseband/vdif/base.py]: I've generalized `_unsqueeze` and am now using it to handle all formats.

File: baseband/vdif/base.py [hunk 16]
```diff
@@ -400,6 +408,12 @@ def _read_frame_set(self, fill_value=0.):
                                             edv=self.header0.edv)
         self._frameset.invalid_data_value = fill_value
 
+    @property
```
[the remark this answers is not in the document]
Comment (cczhu) [reply baseband/vdif/base.py]: `sample_shape` is now a `VLBIStreamBase` property, and can handle squeezing unnamed tuples.

File: baseband/dada/payload.py (hunk 4 above)
Comment (cczhu) [reply baseband/dada/payload.py]: Slight complication - initialization of a `namedtuple` instance is different than that for `tuple`.  I wrote in a try/except block in `VLBIPayloadBase.fromdata` that supports using `namedtuple`, `tuple`, `list` and `np.array` (if a user ever decides to subtype using those as containers).

File: baseband/mark4/payload.py (hunk 7 above)
Comment (cczhu) [reply baseband/mark4/payload.py]: PEP8 agrees, so renamed.

**Comment by cczhu** [comment]
I made changes as suggested by the comments above (I added comments above whenever there were complications during implementation).

I think that having `sample_shape` (squeezed or unsqueezed) handy for end-users beats the inconvenience of not being able to set `squeeze` on the fly when writing arrays to file, so all stream writers now also have a `squeeze` attribute which is set upon initialization, and users now also have access to a (possibly squeezed) sample shape to help guide their file writing.  This also makes the stream reader and writer classes more symmetrical and naturally allows us to move all the squeezing machinery to `VLBIStreamBase`.  The only annoyance is that users must set `fw.squeeze` when they want to change the writing behaviour.

Tests are now included for all formats (except GSB, because I want to overhaul the tests and add docstrings in the next PR) to check their sample shapes in the payloads and

Universalizing squeezing and unsqueezing removes the ability for Mark 4 and 5B stream writers to write a single sample to file without expanding the first dimension, but I can't imagine this being a common operation.

File: baseband/gsb/payload.py [hunk 17]
```diff
@@ -76,6 +77,9 @@ class GSBPayload(VLBIPayloadBase):
                  8: decode_8bit}
     _dtype_word = np.int8
 
+    _sample_shape_cls_1thread = namedtuple('SampleShape', 'nchan')
+    _sample_shape_cls_nthread = namedtuple('SampleShape', 'nthread, nchan')
+
```
Comment (mhvk) [inline baseband/gsb/payload.py]: I think it might make it possible to make the rest more elegant by writing `_sample_shape_cls` as a function:
```
@staticmethod
def _sample_shape_cls(*args):
    if len(args) == 1:
        return _sample_shape_cls_1thread(args[0])
    else:
        return _sample_shape_cls_nthread(*args)
```
I think with that `fromdata` no longer has to be overridden...

File: baseband/gsb/payload.py [hunk 18]
```diff
@@ -94,20 +98,22 @@ def fromfile(cls, fh, payloadsize=None, bps=4, nchan=1,
             Number of bits per sample part (i.e., per channel and per real or
             imaginary component).  Default: 4.
         nchan : int
-            Number of fourier channels.  Default: 1.
+            Number of Fourier channels.  Default: 1.
         complex_data : bool
             Whether data is complex or float.  Default: False.
         """
         if hasattr(fh, 'read'):
+            sample_shape = cls._sample_shape_cls_1thread(nchan)
```
Comment (mhvk) [inline baseband/gsb/payload.py]: If we go with the above, you'd do `sample_shape=(nchan,)` here.

File: baseband/gsb/payload.py [hunk 19]
```diff
@@ -94,20 +98,22 @@ def fromfile(cls, fh, payloadsize=None, bps=4, nchan=1,
...
             return super(GSBPayload,
                          cls).fromfile(fh, payloadsize=payloadsize,
-                                       bps=bps, sample_shape=(nchan,),
+                                       bps=bps, sample_shape=sample_shape,
                                        complex_data=complex_data)
 
         nthread = len(fh)
+        sample_shape_1thread = cls._sample_shape_cls_1thread(nchan)
```
Comment (mhvk) [inline baseband/gsb/payload.py]: and here

File: baseband/gsb/payload.py [hunk 20]
```diff
@@ -126,9 +132,32 @@ def fromfile(cls, fh, payloadsize=None, bps=4, nchan=1,
                 for payload, part in zip(payload_set, thread):
                     part[:] = payload.words.reshape(-1, bpfs // 8)
 
-        return cls(words.ravel(), bps=bps, sample_shape=(nthread, nchan),
+        sample_shape_nthread = cls._sample_shape_cls_nthread(nthread, nchan)
```
Comment (mhvk) [inline baseband/gsb/payload.py]: And of course here `(nthread, nchan)`.

File: baseband/gsb/payload.py [hunk 21]
```diff
@@ -126,9 +132,32 @@ def fromfile(cls, fh, payloadsize=None, bps=4, nchan=1,
...
+        return cls(words.ravel(), bps=bps, sample_shape=sample_shape_nthread,
                    complex_data=complex_data)
 
+    @classmethod
```
Comment (mhvk) [inline baseband/gsb/payload.py]: And with the above, it should no longer be needed to overwrite `fromdata`

File: baseband/mark4/base.py [hunk 22]
```diff
@@ -274,29 +280,30 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
...
-            and squeeze is set to `False`.
+            Array to store the data in. If given, ``count`` will be inferred.
+            If ``squeeze`` is `True`, ``out`` must have squeezed sample
+            dimensions.
 
         Returns
         -------
         out : array of float
-            Dimensions are (sample-time, vlbi-thread, channel).
+            Dimensions are (sample-time, channel) with dimensions
```
Comment (mhvk) [inline baseband/mark4/base.py]: Same comment as for dada.

File: baseband/mark4/base.py [hunk 23]
```diff
@@ -274,29 +280,30 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
...
+            of length unity possibly removed (if ``squeeze`` is `True`).
         """
         if out is None:
             if count is None or count < 0:
                 count = self.size - self.offset
 
-            out = np.empty((self.nthread, count),
-                           dtype=self._frame.dtype).T
+            result = np.empty((self._sample_shape.nchan, count),
```
Comment (mhvk) [inline baseband/mark4/base.py]: Maybe just `self._sample_shape + (count,)` for consistency with dada?

File: baseband/mark4/base.py [hunk 24]
```diff
@@ -342,6 +349,9 @@ class Mark4StreamWriter(VLBIStreamWriterBase, Mark4FileWriter):
         Rate at which each thread is sampled (bandwidth * 2; frequency units).
     header : `~baseband.mark4.Mark4Header`
         Header for the first frame, holding start time information, etc.
+    squeeze : bool, optional
+        If `True` (default), `write` accepts squeezed arrays as input,
```
Comment (mhvk) [inline baseband/mark4/base.py]: double tick on ``write`` (I think, check if sphinx links correctly).

File: baseband/mark4/payload.py [hunk 25]
```diff
@@ -218,11 +221,11 @@ def __init__(self, words, header=None, nchan=1, bps=2, fanout=1):
             self._size = header.payloadsize
         self._dtype_word = MARK4_DTYPES[nchan * bps * fanout]
         self.fanout = fanout
+        sample_shape = self._sample_shape_cls(nchan)
```
Comment (mhvk) [inline baseband/mark4/payload.py]: Doesn't vlbi_base/payload do this already?

File: baseband/mark5b/base.py [hunk 26]
```diff
@@ -198,29 +204,30 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
...
-            and squeeze is set to `False`.
+            Array to store the data in. If given, ``count`` will be inferred.
+            If ``squeeze`` is `True`, ``out`` must have squeezed sample
+            dimensions.
 
         Returns
         -------
         out : array of float or complex
-            Dimensions are (sample-time, vlbi-thread, channel).
+            Dimensions are (sample-time, channel) with dimensions
```
Comment (mhvk) [inline baseband/mark5b/base.py]: As for dada

File: baseband/mark5b/base.py [hunk 27]
```diff
@@ -198,29 +204,30 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
...
+            of length unity possibly removed (if ``squeeze`` is `True`).
         """
         if out is None:
             if count is None or count < 0:
                 count = self.size - self.offset
 
-            out = np.empty((self.nthread, count),
-                           dtype=self._frame.dtype).T
+            result = np.empty((self._sample_shape.nchan, count),
```
Comment (mhvk) [inline baseband/mark5b/base.py]: As above, `self._sample_shape + (count,)`

File: baseband/mark5b/base.py [hunk 28]
```diff
@@ -291,26 +302,40 @@ class Mark5BStreamWriter(VLBIStreamWriterBase, Mark5BFileWriter):
     _frame_class = Mark5BFrame
 
     def __init__(self, raw, frames_per_second=None, sample_rate=None,
-                 nchan=1, bps=2, header=None, **kwargs):
+                 nchan=1, bps=2, header=None, squeeze=True, **kwargs):
         if header is None:
             header = Mark5BHeader.fromvalues(**kwargs)
+        sample_shape = Mark5BPayload._sample_shape_cls(nchan)
```
Comment (mhvk) [inline baseband/mark5b/base.py]: Seeing this time and again, I wonder whether `vlbi_base/base.py` should simply use `self._payload_cls._sample_shape_cls(...)` but let's leave that for possible follow up.

File: baseband/mark5b/payload.py [hunk 29]
```diff
@@ -109,14 +110,16 @@ class Mark5BPayload(VLBIPayloadBase):
     _encoders = {2: encode_2bit}
     _decoders = {2: decode_2bit}
 
+    _sample_shape_cls = namedtuple('SampleShape', 'nchan')
+
     def __init__(self, words, nchan=1, bps=2, complex_data=False):
         if complex_data:
             raise ValueError("Mark5B format does not support complex data.")
 
+        sample_shape = self._sample_shape_cls(nchan)
```
Comment (mhvk) [inline baseband/mark5b/payload.py]: As for mark4, I'd think that `vlbi_base` could do this call.

File: baseband/vdif/base.py [hunk 30]
```diff
@@ -344,29 +354,31 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
...
-            and squeeze is set to `False`.
+            Array to store the data in. If given, ``count`` will be inferred.
+            If ``squeeze`` is `True`, ``out`` must have squeezed sample
+            dimensions.
 
         Returns
         -------
         out : array of float or complex
-            Dimensions are (sample-time, vlbi-thread, channel).
+            Dimensions are (sample-time, vlbi-thread, channel) with dimensions
```
Comment (mhvk) [inline baseband/vdif/base.py]: As for dada.

File: baseband/vdif/payload.py [hunk 31]
```diff
@@ -143,8 +146,9 @@ def __init__(self, words, header=None,
                 self._encoders = Mark5BPayload._encoders
                 if complex_data:
                     raise ValueError("VDIF/Mark5B payload cannot be complex.")
+        sample_shape = self._sample_shape_cls(nchan)
```
Comment (mhvk) [inline baseband/vdif/payload.py]: Move to `vlbi_base`?

File: baseband/vlbi_base/base.py [hunk 32]
```diff
@@ -70,6 +71,31 @@ def __init__(self, fh_raw, header0, nchan, bps, complex_data, thread_ids,
 
         self.frames_per_second = frames_per_second
         self.offset = 0
+        self.squeeze = squeeze
+
+    @property
+    def sample_shape(self):
+        """Shape of a data sample (possibly squeezed).
+        """
```
Comment (mhvk) [inline baseband/vlbi_base/base.py]: Move closing `"""` to previous line for single-line docstrings.

File: baseband/vlbi_base/base.py [hunk 33]
```diff
@@ -70,6 +71,31 @@ def __init__(self, fh_raw, header0, nchan, bps, complex_data, thread_ids,
...
+        if self.squeeze:
+            if not self._squeezed_shape:
+                dims = np.array(self._sample_shape)
```
Comment (mhvk) [inline baseband/vlbi_base/base.py]: I think going via `np.array` is needlessly expensive. Maybe just (note: didn't check this, so may not work immediately)
```
field_names = getattr(self._sample_shape, '_fields', None)
if field_names is not None:
    squeezed_shape_cls = namedtuple('SampleShape', 
            ','.join(*[field_name for field_name, dim in zip(field_names, self._sample_shape) if dim > 1]))
else:
    squeezed_shape_cls = tuple
self._squeezed_shape = squeezed_shape_cls(*[dim for dim in self._sample_shape if dim > 1])
```

File: baseband/vlbi_base/base.py [hunk 34]
```diff
@@ -70,6 +71,31 @@ def __init__(self, fh_raw, header0, nchan, bps, complex_data, thread_ids,
...
+                except AttributeError:
+                    self._squeezed_shape = tuple(dims[keep_dims])
+                else:
+                    sqz_shp = namedtuple('SampleShape', field_names[keep_dims])
+                    self._squeezed_shape = sqz_shp(*dims[keep_dims])
+            return self._squeezed_shape
+        return self._sample_shape
+
+    def _unsqueeze(self, data):
+        unity_dims = np.where(np.array(self._sample_shape) == 1)[0] + 1
```
Comment (mhvk) [inline baseband/vlbi_base/base.py]: This can be done more directly:
```
new_shape = list(data.shape)
for i, dim in enumerate(self._sample_shape):
    if dim == 1:
        new_shape.insert(i, 1)  # maybe i+1?
return data.reshape(new_shape)
```
Actually: just to be sure, this loop may have to go backwards (in case there is more than one unity dimension). Do write some tests directly testing this logic!

File: baseband/vlbi_base/payload.py [hunk 35]
```diff
@@ -98,7 +100,12 @@ def fromdata(cls, data, bps=2):
             Number of bits per sample to use (for complex data, for real and
             imaginary part separately; default: 2).
         """
-        sample_shape = data.shape[1:]
+        try:
+            sample_shape = cls._sample_shape_cls(*data.shape[1:])
```
Comment (mhvk) [inline baseband/vlbi_base/payload.py]: This is why I thought you didn't need to initialize using `_sample_shape_cls` in the subclasses, but I now see it is just for `fromdata`. If we move this to `__init__`, it may not be necessary to do anything special here.

File: baseband/vlbi_base/payload.py [hunk 36]
```diff
@@ -37,12 +37,14 @@ class VLBIPayloadBase(object):
...
     _dtype_word = np.dtype('<u4')
     """Default for words: 32-bit unsigned integers, with lsb first."""
     # To be defined by subclasses.
     _encoders = {}
     _decoders = {}
+    # Placeholder for sample shape named tuple.
+    _sample_shape_cls = tuple
 
     def __init__(self, words, bps=2, sample_shape=(), complex_data=False):
         self.words = words
```
Comment (mhvk) [inline baseband/vlbi_base/payload.py]: I think we might as well pass `sample_shape` through `self._sample_shape_cls` here.

File: docs/baseband/mark4/index.rst [hunk 37]
```diff
@@ -39,7 +39,8 @@ can be calculated (for longer files, this can be calculated from the file)::
     ...                 frames_per_second=400)
     >>> fh
     <Mark4StreamReader name=... offset=0
-        samples_per_frame=80000, nchan=8, frames_per_second=400, bps=2,
+        frames_per_second=400, samples_per_frame=80000,
+        sample_shape=SampleShape(nchan=8), bps=2,
```
Comment (mhvk) [inline docs/baseband/mark4/index.rst]: The example makes me wonder whether we should call these `Mark4Shape`, etc... (as users cannot initialize these themselves...) But maybe leave that for another PR.

File: baseband/dada/base.py [hunk 38]
```diff
@@ -242,40 +235,38 @@ def header1(self):
...
+            dimensions.
 
         Returns
         -------
         out : array of float or complex
-            Dimensions are (sample-time, thread, channel), with dimensions of
-            length unity possibly removed (if ``squeeze`` is `True`, or if they
-            were not present in the ``out`` array passed in).
+            Dimensions are (sample-time, thread (polarization), channel), with
+            dimensions of length unity possibly removed (if ``squeeze`` is
```
[the remark this answers is not in the document]
Comment (cczhu) [reply baseband/dada/base.py]: Changed.

File: baseband/gsb/base.py [hunk 39]
```diff
@@ -197,11 +229,8 @@ def read(self, count=None, fill_value=0., squeeze=True, out=None):
...
             Value to use for invalid or missing data.
-        squeeze : bool
-            If `True` (default), remove channel and thread dimensions if unity.
         out : `None` or array
-            Array to store the data in. If given, count will be inferred,
-            and squeeze is set to `False`.
+            Array to store the data in. If given, count will be inferred.
 
         Returns
         -------
```
[the remark this answers is not in the document]
Comment (cczhu) [reply baseband/gsb/base.py]: Changed to match DADA's.

File: baseband/gsb/payload.py (hunk 17 above)
Comment (cczhu) [reply baseband/gsb/payload.py]: So it does!  I've updated `GSBPayload` and added a few lines to `test_gsb.py` to ensure we're still getting the correct sample shapes.

Made corresponding changes below.

File: baseband/gsb/payload.py (hunk 18 above)
Comment (cczhu) [reply baseband/gsb/payload.py]: Changed.

File: baseband/gsb/payload.py (hunk 19 above)
Comment (cczhu) [reply baseband/gsb/payload.py]: Changed.

File: baseband/gsb/payload.py (hunk 20 above)
Comment (cczhu) [reply baseband/gsb/payload.py]: Changed.

File: baseband/gsb/payload.py (hunk 21 above)
Comment (cczhu) [reply baseband/gsb/payload.py]: Changed as per above.

File: baseband/mark4/base.py (hunk 22 above)
Comment (cczhu) [reply baseband/mark4/base.py]: Changed.

File: baseband/mark4/base.py (hunk 23 above)
Comment (cczhu) [reply baseband/mark4/base.py]: It also looks nicer and is shorter, so changed for all formats except VDIF.

File: baseband/mark4/base.py (hunk 24 above)
Comment (cczhu) [reply baseband/mark4/base.py]: Changed.

File: baseband/mark4/payload.py (hunk 25 above)
Comment (cczhu) [reply baseband/mark4/payload.py]: No, it's a `tuple` in `vlbi_base`.  If we declare it in `VLBIPayloadBase`, then any end-user who creates a new format will have to declare a proper `sample_shape`.  I recommend leaving this for when we decide whether to implement `self._payload_cls._sample_shape_cls`.

File: baseband/mark5b/base.py (hunk 26 above)
Comment (cczhu) [reply baseband/mark5b/base.py]: Changed.

File: baseband/mark5b/base.py (hunk 27 above)
Comment (cczhu) [reply baseband/mark5b/base.py]: Changed.

File: baseband/vdif/base.py (hunk 30 above)
Comment (cczhu) [reply baseband/vdif/base.py]: Changed.

File: baseband/vlbi_base/base.py (hunk 32 above)
Comment (cczhu) [reply baseband/vlbi_base/base.py]: Changed.

File: baseband/vlbi_base/base.py (hunk 33 above)
Comment (cczhu) [reply baseband/vlbi_base/base.py]: Running a few `timeit` tests, using arrays is only slightly more computationally expensive than list comprehension.  I switched to list comprehension, and agree it's more reasonable to call lower-level machinery here.

File: baseband/vlbi_base/base.py (hunk 34 above)
Comment (cczhu) [reply baseband/vlbi_base/base.py]: I agree that's more direct and doesn't require higher-level machinery (though I liked how compact mine was).  Changed, and included tests in `test_vlbi_base.py`.

File: baseband/vlbi_base/payload.py (hunk 35 above)
Comment (cczhu) [reply baseband/vlbi_base/payload.py]: I'm not sure I understand - are you recommending that we move the try/except block to `__init__` (and pass `data.shape[1:]` as a tuple to it in `fromdata`).  I'm not sure that makes sense since the try/except is never used except in `fromdata`, and sample shapes are already properly initialized in format subclasses.

File: docs/baseband/mark4/index.rst (hunk 37 above)
Comment (cczhu) [reply docs/baseband/mark4/index.rst]: I played with the idea, but didn't like it because there's nothing about the sample shape that makes it unique to Mark 4.  Even for a single file, the shape could be different between squeezed and non-squeezed.

**Comment by cczhu** [comment]
Changes made as per comments.  I'm not sure I fully agree with moving the named tuple `sample_shape` up to `VLBIStreamBase`, as tuples are able to take in any number of values, but this number is hard-coded for sample shapes.  I've also not switched sample shape class calls in format payload classes to `self._payload_cls._sample_shape_cls`.  I think that both, if we agree they should be done, can be handled in the next PR.  Right now I'd just like to merge our current changes so I can merge with the Getting Started docs and update the latter to reflect these changes.

File: baseband/mark4/payload.py (hunk 25 above)
Comment (mhvk) [reply baseband/mark4/payload.py]: I don't think it is true any end user would have to declare anything!

File: baseband/vlbi_base/payload.py [hunk 40]
```diff
@@ -37,12 +37,14 @@ class VLBIPayloadBase(object):
...
     _size = None
-    # Default type for encoded data
+    # Default type for encoded data.
     _dtype_word = np.dtype('<u4')
     """Default for words: 32-bit unsigned integers, with lsb first."""
     # To be defined by subclasses.
     _encoders = {}
     _decoders = {}
+    # Placeholder for sample shape named tuple.
+    _sample_shape_cls = tuple
```
Comment (mhvk) [inline baseband/vlbi_base/payload.py]: Looking at this again, I realize `tuple` is not really a good analogy to a namedtuple instance, since one cannot do `tuple(*a)` (which also leads to the try/except in `fromdata`). I think we should write instead:
```
# outside of class
def _sample_shape_tuple(*sample_shape):
    return sample_shape
```

_sample_shape_cls = _sample_shape_tuple
```
(the construct of `*sample_shape` is guaranteed to be turned into a tuple already). Alternatively, you could directly define the function here.

File: baseband/vlbi_base/payload.py (hunk 36 above)
Comment (mhvk) [reply baseband/vlbi_base/payload.py]: So, I'd still argue to here do
```
self.sample_shape = self._sample_shape_cls(*sample_shape)
```
(with the above definition, this will always work)

File: baseband/vlbi_base/payload.py (hunk 35 above)
Comment (mhvk) [reply baseband/vlbi_base/payload.py]: With the above, passing `data.shape[1:]` will work fine.

File: baseband/vlbi_base/payload.py (hunk 40 above)
Comment (mhvk) [reply baseband/vlbi_base/payload.py]: Maybe even easier would be to do
```
_sample_shape_cls = None
```

and then below
```
if self._sample_shape_cls is not None:
    sample_shape = self._sample_shape_cls(*sample_shape)
```

Comment (mhvk) [reply baseband/vlbi_base/payload.py]: Rename as `_sample_shape_maker`?

**Comment by cczhu** [comment]
Applied all recommended changes and rewrote the docstring.  The only place where `squeeze=True` is still mentioned is in the Returns section of `read`, since squeezing a single sample will also remove the sample-time dimension.

**Comment by mhvk** [comment]
Thanks! Merged...
